### PR TITLE
[mqtt] Remove failure crate dependency

### PR DESCRIPTION
**Pull request by dmolokanov**
The project uses `failure` crate to preserve the context of the error. While it works fine it seems like maintainer is planning to discontinue it soon. Another issue is that the crate is based on custom `Fail` trait which is not compatible with `std::error::Error` trait.
The idea is to get rid of `failure` crate in favor of `thiserror` which seems to provide almost everything we need and based on `std::error::Error`.

### CODE REVIEW

File: mqtt/mqtt-broker/src/auth/authentication.rs [hunk 1]
```diff
@@ -13,38 +12,56 @@ pub enum Credentials {
...
+    fn from(certificate: Vec<u8>) -> Self {
+        Self(certificate)
+    }
+}
+
 /// A trait to authenticate a MQTT client with given credentials.
 #[async_trait]
 pub trait Authenticator {
+    /// Authentication error.
+    type Error;
```
Comment (myagley) [inline mqtt/mqtt-broker/src/auth/authentication.rs:28]: should we force this to be `std::error::Error`? The broker below has a type bound on `Display` which is so close to being `std::error::Error` (just need to add `source`)

Comment (myagley) [reply mqtt/mqtt-broker/src/auth/authentication.rs:28]:
```rust
type Error: std::error::Error;
```

File: mqtt/mqtt-broker/src/broker.rs [hunk 2]
```diff
@@ -32,7 +30,9 @@ macro_rules! try_send {
 pub struct Broker<N, Z>
 where
     N: Authenticator,
+    N::Error: Display + Into<Error>,
     Z: Authorizer,
+    Z::Error: Display + Into<Error>,
```
Comment (myagley) [inline mqtt/mqtt-broker/src/broker.rs:35]: We only use this error for logging? Do we ever bubble this error out or take action on it?

File: mqtt/mqtt-broker/src/error.rs [hunk 3]
```diff
@@ -1,135 +1,88 @@
...
+    SendConnectionMessage(#[source] tokio::sync::mpsc::error::SendError<Message>),
 
-    #[fail(display = "An error occurred sending a message to a snapshotter.")]
+    #[error("An error occurred sending a message to a snapshotter.")]
     SendSnapshotMessage,
 
-    #[fail(display = "An error occurred decoding a packet.")]
-    DecodePacket,
+    #[error("An error occurred decoding a packet.")]
+    DecodePacket(#[from] mqtt3::proto::DecodeError),
```
Comment (myagley) [inline mqtt/mqtt-broker/src/error.rs:20]: Does `#[from]` also set the `#[source]`?

File: mqtt/mqttd/src/main.rs [hunk 4]
```diff
@@ -1,7 +1,6 @@
 use std::{convert::TryInto, env, io};
```
Comment (myagley) [inline mqtt/mqttd/src/main.rs:1]: Just a general note, we should probably keep an eye on the length of `main`. I don't want what we have in `iotedged` to happen again where the `main` method is huge and not tested. Nothing required to change in this PR. Edits to `main` reminded me.

File: mqtt/mqtt-broker/src/error.rs (hunk 3 above)
Comment (dmolokanov) [reply mqtt/mqtt-broker/src/error.rs:20]: yes, but not vice versa. `#[source]` dont generate `impl From`

Comment (dmolokanov) [reply mqtt/mqtt-broker/src/error.rs:20]: > The #[from] attribute always implies that the same field is #[source], so you don't ever need to specify both attributes.

**Pull request merged**